### PR TITLE
Add unit test on duplicate batch for Espresso streamer

### PR DESCRIPTION
**Pull request by dailinsubjam**
Closes https://app.asana.com/1/1208976916964769/project/1209393353274209/task/1210053042826958?focus=true

### This PR:
- Adds a unit test for espresso streamer on the case of duplicated batch (= duplicated espresso transaction)
    - It uses existing testing infrastructure for the streamer
    - It expects the streamer to skip the duplicated batch

### Key places to review:
`espresso/streamer_test.go`

### How to test this PR: 
`go test -run ^TestEspressoStreamerDuplicationHandling$ ./espresso`
Now the test passes in 0.77s

### CODE REVIEW

File: espresso/streamer_test.go [hunk 1]
```diff
@@ -610,3 +610,79 @@ func TestStreamerEspressoOutOfOrder(t *testing.T) {
 		t.Fatalf("unexpected number of batches in state:\nhave:\n\t%v\nwant:\n\t%v\n", have, want)
 	}
 }
+
+// TestEspressoStreamerDuplicationHandling tests the behavior of the EspressoStreamer
+// when the duplicat batch is received.
```
Comment (shenkeyao) [inline espresso/streamer_test.go]:
```suggestion
// when a duplicated batch is received.
```

File: espresso/streamer_test.go [hunk 2]
```diff
@@ -610,3 +610,79 @@ func TestStreamerEspressoOutOfOrder(t *testing.T) {
...
+
+	// update the state of our streamer
+	syncStatus := state.SyncStatus()
+	_, err := streamer.Refresh(ctx, syncStatus.FinalizedL1, syncStatus.SafeL2.Number, syncStatus.SafeL2.L1Origin)
+
+	if have, want := err, error(nil); have != want {
+		t.Fatalf("failed to refresh streamer state encountered error:\nhave:\n\t\"%v\"\nwant:\n\t\"%v\"\n", have, want)
+	}
+
+	const N = 1000
```
Comment (shenkeyao) [inline espresso/streamer_test.go:639]: I think it's okay to test with a smaller `N`, such as `10`, since this isn't a stress test. It's not a big deal though, if this test doesn't take too long.

File: packages/contracts-bedrock/lib/forge-std
Comment (shenkeyao) [inline packages/contracts-bedrock/lib/forge-std:1]: Have you synced with Philippe's fix for the contract build? Maybe we should remove this file because it's not unrelated to the test.

File: packages/contracts-bedrock/lib/superchain-registry
Comment (shenkeyao) [inline packages/contracts-bedrock/lib/superchain-registry:1]: Similarly, I was wondering whether we should include this file in the PR.

File: espresso/streamer_test.go [hunk 3]
```diff
@@ -610,3 +610,79 @@ func TestStreamerEspressoOutOfOrder(t *testing.T) {
...
+			t.Fatalf("batch epoch number does not match:\nhave:\n\t%v\ndo not want:\n\t%v\n", have, want)
+		}
+
+		state.AdvanceSafeL2()
+		state.AdvanceFinalizedL1()
+
+	}
+
+	// Check that the state has the correct number of duplicated batches
+	if have, want := len(state.EspTransactionData), 2*N; have != want {
```
Comment (QuentinI) [inline espresso/streamer_test.go]: Here as well

File: espresso/streamer_test.go [hunk 4]
```diff
@@ -610,3 +610,79 @@ func TestStreamerEspressoOutOfOrder(t *testing.T) {
...
+				t.Fatalf("failed to update streamer state encountered error:\nhave:\n\t\"%v\"\nwant:\n\t\"%v\"\n", have, want)
+			}
+		}
+
+		batchFromEsp := streamer.Next(ctx)
+		require.NotNil(t, batchFromEsp, "unexpectedly did not receive a batch from streamer")
+
+		// This batch ** should ** match the one we created above.
+		// If the duplicate one is NOT skipped, this will FAIL.
+		if have, want := batchFromEsp.Batch.GetEpochNum(), batch.GetEpochNum(); have != want {
```
Comment (QuentinI) [inline espresso/streamer_test.go]: Can we use require.Equal here instead?

File: espresso/streamer_test.go [hunk 5]
```diff
@@ -610,3 +610,79 @@ func TestStreamerEspressoOutOfOrder(t *testing.T) {
...
+			chainSigner,
+		)
+
+		// duplicate the batch
+		for j := 0; j < 2; j++ {
+			// update the state of our streamer
+			syncStatus := state.SyncStatus()
+			_, err := streamer.Refresh(ctx, syncStatus.FinalizedL1, syncStatus.SafeL2.Number, syncStatus.SafeL2.L1Origin)
+
+			if have, want := err, error(nil); have != want {
```
Comment (QuentinI) [inline espresso/streamer_test.go]: Can we replace this and following with require.NoError?

**Review by QuentinI — COMMENTED** [review]
Some nitpicks, LGTM overall

File: packages/contracts-bedrock/lib/forge-std
Comment (philippecamacho) [reply packages/contracts-bedrock/lib/forge-std:1]: @shenkeyao I did the rebase and now the diff only contains the test file.

File: packages/contracts-bedrock/lib/superchain-registry
Comment (philippecamacho) [reply packages/contracts-bedrock/lib/superchain-registry:1]: See https://github.com/EspressoSystems/optimism-espresso-integration/pull/153#discussion_r2114123435

File: espresso/streamer_test.go (hunk 2 above)
Comment (dailinsubjam) [reply espresso/streamer_test.go:639]: Right the test is pretty fast (< 1s), so I just keep N as 1000 to be same as other unit test settings.

File: espresso/streamer_test.go (hunk 3 above)
Comment (dailinsubjam) [reply espresso/streamer_test.go]: All updated!

**Review by QuentinI — APPROVED** [review]
